shared/src/main/scala/org/widok/validation/Validation.scala

```diff
@@ -0,0 +1,142 @@
+package org.widok.validation
+
+abstract class FieldValidation[T](params: Map[String, Any] = Map.empty, message: String = "validation failed") {
+
+  /**
+   * Implement this method in subclasses to validate a value.
+   *
+   * @param value The value to be validated
+   * @return Either a boolean indicating if validation is successful or an Optional string with a
+   *         validation error message. If a boolean (false) is returned an error message will be constructed using the
+   *         formatMessage method.
+   */
+  protected def validate(value: T): Either[Boolean, Option[String]]
+
+  def validateValue(value: Any): Option[String] = {
+    val typedValue: T = value.asInstanceOf[T]
+
+    validate(typedValue) match {
+      case Left(result) => if (result) None else Some(formatMessage(message, params, typedValue))
+      case Right(validationMsg) => validationMsg
+    }
+  }
+
+  def formatMessage(msg: String, params: Map[String, Any], value: T) = {
+    val paramsWithValue = params + ("value" -> value)
+    (msg /: paramsWithValue) { (res, entry) => res.replaceAll("#\\{%s\\}".format(entry._1), entry._2.toString.replace( """\""", """\\""")) }
+  }
+}
+
+abstract class TextFieldValidation(params: Map[String, Any] = Map.empty, message: String = "validation failed") extends FieldValidation[String](params, message) {
+  override def validateValue(value: Any): Option[String] = {
+    if (!value.isInstanceOf[String]) throw new IllegalArgumentException("Expected a string value")
+    val typedValue = value.asInstanceOf[String]
+    if (typedValue.isEmpty) None else super.validateValue(typedValue)
+  }
+}
+
+object Validations {
+
+  /**
+   * Validates min values
+   *
+   * @param min The minimum value for the field
+   * @param message Error message. Variables: #{value} the validated input, #{min} = the min value
+   */
+  case class MinValidation(min: Int, message: Option[String] = None)
+    extends TextFieldValidation(Map("min" -> min), message.getOrElse("Value must be at least #{min}")) {
```

Comment:
Just to clearify. The idea is that each Validation defines a Map of specific variables for that Validation that can be used when formatting error messages #{value} is the only one that is available in all Validations. So, the formatting will take place in the ```FieldValidation```class, using the Map of variables and message that is passed in through the constructor in line 47. The actual interpolation code is a bit of a mess but I didn't find any clean way of doing this. 

```scala
  def formatMessage(msg: String, params: Map[String, Any], value: T) = {
    val paramsWithValue = params + ("value" -> value)
    (msg /: paramsWithValue) { (res, entry) => res.replaceAll("#\\{%s\\}".format(entry._1), entry._2.toString.replace( """\""", """\\""")) }
  }
```
http://stackoverflow.com/questions/4051308/better-string-formatting-in-scala

This allows the application to customize error messages, like this:

```scala
new MinValidation(10, "You entered %{value} but the min acceptable number is: %{min}")
```